Add create timeline page

leenagupte opened the pull request:
Trello: https://trello.com/c/Gqe4lgaI

# What?

Create a page to allow new timeline entries to be stored in the database. 

Changes the markdown editor component to allow users to see the full range of menu options, whilst keeping the existing menu restrictions on the "step" and "sub-section" edit pages.

Only users with "Unreleased feature" permissions should be able to access the page whilst the timeline feature is being built.

The timeline entry is only persisted to the database. Whilst we are calling `DraftUpdater`, to send changes to the publishing-api, the actual creation of the payload for publishing-api will happen in another story.

# Expected changes

URL of page: http://collections-publisher.dev.gov.uk/coronavirus/landing/timeline_entries/new

![screenshot-collections-publisher dev gov uk-2021 01 12-10_15_33](https://user-images.githubusercontent.com/5793815/104301441-8c0bc380-54bf-11eb-9944-953fbc1d3fb6.png)

Error messages:

<img width="1532" alt="Screenshot 2021-01-13 at 17 15 59" src="https://user-images.githubusercontent.com/5793815/104489265-56ea9880-55c7-11eb-86d0-b63dea6bbe14.png">


⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

kevindew commented on app/views/timeline_entries/_form.html.erb:
```diff
@@ -0,0 +1,33 @@
+  blockquote_button: true,
+  numbered_list_button: true,
+  bullets_button: true
+} %>
```
It looks like this is missing the error display that is shown for the heading field.

For example: 

![Screenshot 2021-01-12 at 11 06 25](https://user-images.githubusercontent.com/282717/104306764-40a8e380-54c6-11eb-8d73-2d95a57847ad.png)

leenagupte replied:
Ah! It looks like this is missing in the component. I'll add it.

kevindew commented on app/views/timeline_entries/new.html.erb:
```diff
@@ -1 +1,35 @@
+  metadata = {
+    "Status" => "Draft",
+    "Last saved" => format_full_date_and_time(DateTime.now),
+  }
```
It doesn't look like this metadata variable is used - unless I'm missing something?

leenagupte replied:
I think you're right. I'll remove it.

kevindew commented on app/views/timeline_entries/new.html.erb:
```diff
@@ -1 +1,35 @@
+
+<div class="covid-edit-page govuk-grid-row">
+  <div class="govuk-grid-column-two-thirds">
+    <%= form_for @timeline_entry, announcement: @timeline_entry, url: coronavirus_page_timeline_entries_path do  %>
```
I don't think this "announcement" argument does anything, is it something caught in a copy & paste from announcements?

kevindew replied:
I did wonder if you need the url and timeline_entry referenced here and discovered a TIL in the Rails manual.

It looks like `form_for` is soft deprecated and the conventional Rails way would be to use [`form_with`](https://edgeguides.rubyonrails.org/form_helpers.html). So this should probably be `<%= form_with url: coronavirus_page_timeline_entries_path do %>`

leenagupte replied:
You caught me with the copy and paste error. Oops.

leenagupte replied:
Hmmm. The form stops submitting when I switch to a `form_with`.

leenagupte replied:
Ah! Figured it out. I need to set `local: true`

kevindew replied:
Oh gosh `local: true` sounds a strange attribute 🤔

kevindew commented on app/controllers/timeline_entries_controller.rb:
```diff
@@ -7,9 +7,27 @@ def new
+    if @timeline_entry.save && draft_updater.send
+      redirect_to coronavirus_page_path(coronavirus_page.slug), notice: "Timeline entry was successfully created."
+    else
+      render :new
```
this should probably have a status of unprocessable entity to indicate that validation failed - but maybe don't bother if it's a nightmare for consistency with the rest of the app.

kevindew reviewed (commented):
Looking great 👍  I've got a few minor comments here, nothing too significant to resolve.

leenagupte commented on spec/components/markdown_editor_spec.rb:
```diff
@@ -22,9 +22,9 @@
+  it "renders bullet list toolbar button if bullets_button is true" do
     render "components/markdown_editor",
-           hide_bullets_button: true,
+           bullets_button: true,
```
Yeah, I didn't know about it either until I tried to make this change.

leenagupte commented on spec/controllers/timeline_entries_controller_spec.rb:
```diff
@@ -0,0 +1,26 @@
+      get :new, params: { coronavirus_page_slug: coronavirus_page.slug }
+      expect(response).to have_http_status(:forbidden)
+    end
+  end
```
I'll change this to set the permissions explicitly.

leenagupte commented on app/views/components/_markdown_editor.html.erb:
```diff
@@ -3,6 +3,7 @@
   component_classes = %w[app-c-markdown-editor govuk-form-group]
   heading_buttons ||= false
   blockquote_button ||= false
+  numbered_list_button ||= false
```
My initial thought was to have all of the controls turned on by default, but I couldn't get it to work with the default set to "true", the options were always on. That's why I went with this approach of having to turn on the toolbar options you want.

I used the booleans approach because that seemed to follow the existing pattern for components.  So are you suggesting, having say `controls ||= [:headings, :blockquotes]` set here, so in this example, if you don't pass in controls, you'll see the heading and blockquotes buttons, but that if you want to see more, you have to explicitly pass them in? e.g. `controls: [:headings, :blockquotes, :numbered_list]`?

kevindew replied:
Yup, that sounds a good approach to me. Can do that in a separate PR if you like to not block this one?

leenagupte replied:
I've got a separate commit ready. Will push that first to check everything is alright before splitting it up and fixing up.

leenagupte commented:
@kevindew Thanks for your review. I've addressed all of your comments. As mentioned in one of the replies, the change to passing a list of optional controls to the markdown editor is in a separate commit to check the approach is ok before splitting it up.

I've also switched to using `form_with` and tried to explain in the commit message why `local` needs to be set to true.

I've also added a feature test that I forgot to add the first time. Oops.

If get a chance can you re-review?

kevindew commented on app/views/timeline_entries/new.html.erb:
```diff
@@ -0,0 +1,32 @@
+
+<div class="covid-edit-page govuk-grid-row">
+  <div class="govuk-grid-column-two-thirds">
+    <%= form_with model: @timeline_entry,
```
I don't think this model does anything - since we reference the [model directly in the form](https://github.com/alphagov/collections-publisher/pull/1235/files#diff-0ed636e7df998d6718cbb06b6a880f21c098ce3020e9bed952165b9018b01794R21). So I think we can remove this without consequences.

leenagupte replied:
Ah, yes, I think this was one of the many options I tried to get the form to submit before stumbling on the `local` issue. I'll remove it.

kevindew commented on app/views/timeline_entries/new.html.erb:
```diff
@@ -0,0 +1,32 @@
+  <div class="govuk-grid-column-two-thirds">
+    <%= form_with model: @timeline_entry,
+                  url: coronavirus_page_timeline_entries_path,
+                  local: true do %>
```
I did some reading about this, and wow what a super annoying quirk! The good news is that Rails recognised this is weird too and in [Rails 6.1 this is going to default to be a local form](https://github.com/rails/rails/pull/40946) and you have to specify remote to change that.

Rather than specifying `local` here on this form I'd suggest we change config/application.rb to have `config.action_view.form_with_generates_remote_forms = false` so that this configuration is applied globally rather than needed on individual forms. We can leave a note with that to say we can remove that line once we've adopted Rails 6.1 config.

leenagupte replied:
This is much nicer! I'll add a comment to config explaining why it's there.

kevindew commented on app/views/timeline_entries/_form.html.erb:
```diff
@@ -0,0 +1,31 @@
+    value: @timeline_entry.content
+  },
+  error_message: error_items(@timeline_entry.errors.messages, :content),
+  controls: [:headings, :blockquote, :numbered_list, :bullets]
```
Yay - I like this 👍

kevindew reviewed (commented):
Looking great Leena, just a couple of things that I think we can tweak with the form_with

leenagupte commented:
@kevindew Thanks for your comments. I've addressed them and fixed up. Could you re-re-review?

kevindew reviewed (approved):
Nice one 👍
